### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Projectivization): cardinality over finite fields

### PR DESCRIPTION
**Pull request by chrisflav**
We compute the cardinality of the projectivization of a finite vector space over a finite field.

From "Formalizing the Bruhat-Tits tree"

Co-authored-by: Judith Ludwig <ludwig@mathi.uni-heidelberg.de>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/Finite/Basic.lean [hunk 1]
```diff
@@ -249,6 +249,15 @@ theorem card' : ∃ (p : ℕ) (n : ℕ+), Nat.Prime p ∧ Fintype.card K = p ^ (
...
   ⟨p, @FiniteField.card K _ _ p hc⟩
 
+omit [Fintype K] in
+lemma two_le_card [Finite K] : 2 ≤ Nat.card K := by
+  haveI : Fintype K := Fintype.ofFinite K
+  obtain ⟨p, n, h⟩ := FiniteField.card' K
+  rw [← Fintype.card_eq_nat_card, h.right, ← pow_one 2]
+  exact le_trans
+    (Nat.pow_le_pow_of_le (by simp only [Nat.one_lt_ofNat]) NeZero.one_le)
+    (Nat.pow_le_pow_of_le_left (Nat.Prime.two_le h.left) n)
```
Comment (acmepjz) [inline Mathlib/FieldTheory/Finite/Basic.lean:259]: (untested)

```suggestion
lemma two_le_card [Finite K] : 2 ≤ Nat.card K := by
  Finite.one_lt_card
```

This should work for any type which is `Finite` and `Nontrivial` (which automatically holds for `Field`). Maybe you want to add `Finite.two_le_card` instead.

Comment (chrisflav) [reply Mathlib/FieldTheory/Finite/Basic.lean]: Indeed, this lemma is just nonsense, I don't know what I was thinking. Thanks!

File: Mathlib/LinearAlgebra/Projectivization/Cardinality.lean [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  e.trans (Equiv.prodCongrLeft (fun _ ↦ (equivQuotientOrbitRel k V).symm))
+
+/-- If `V` is a finite `k`-module and `k` is finite, `ℙ k V` is finite. -/
+instance finite_of_finite [Finite V] : Finite (ℙ k V) :=
+  have : Finite (ℙ k V × kˣ) := Finite.of_equiv _ (nonZeroEquivProjectivizationProdUnits k V)
+  Finite.prod_left kˣ
+
+/-- Fraction free cardinality formula for the points of `ℙ k V` if `k` and `V` are finite.
+See `Projectivization.card'` and `Projectivization.card''` for other spellings of the formula. -/
+lemma card [Finite k] [Finite V] :
```
[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/LinearAlgebra/Projectivization/Cardinality.lean:41]: Or do we want an infinite version? I think if `Infinite k`, then `#V = #(ℙ k V) * #k` (modulo some `Cardinal.lift`).

[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/LinearAlgebra/Projectivization/Cardinality.lean]: I think there is no real gain of having this over `nonZeroEquivProjectivizationProdUnits`.

[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/LinearAlgebra/Projectivization/Cardinality.lean]: > I think there is no real gain of having this over `nonZeroEquivProjectivizationProdUnits`?

Indeed. Feel free to ignore my blindness suggestion...

File: Mathlib/LinearAlgebra/Projectivization/Cardinality.lean [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+as a fraction. -/
+lemma card'' [Finite k] : Nat.card (ℙ k V) = (Nat.card V - 1) / (Nat.card k - 1) := by
+  haveI : Fintype k := Fintype.ofFinite k
+  rw [card k]
+  have : 1 < Nat.card k := Finite.one_lt_card
+  have h : 0 ≠ (Nat.card k - 1) := by omega
+  exact Nat.eq_div_of_mul_eq_left (Ne.symm h) rfl
+
+lemma card_of_finrank_two [Finite k] (h : Module.finrank k V = 2) :
+    Nat.card (ℙ k V) = Nat.card k + 1 := by
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Projectivization/Cardinality.lean:97]: Could you please also add the generalization $\sum_{i < r} q^i$, where `r` is the rank and `q` is the cardinality of `k`?

Comment (chrisflav) [reply Mathlib/LinearAlgebra/Projectivization/Cardinality.lean:120]: Done, thanks for the suggestion

**Review by jcommelin — COMMENTED** [review]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors  merge

**Comment by chrisflav** [comment]
I merged master, @riccardobrasca could you please give this one another `bors merge`?

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**